File: BOJ/python3/15655.py
'''
N과 M (6)
https://www.acmicpc.net/problem/15655
'''
def dfs(lv, cur_n): # lv: 트리의 레벨, cur_n: 현재 n
    if lv == M:
        for n in p:
            print(n, end=' ')
        print()
        return
    else:
        if lv == 0:
            cur_n_idx = -1
        else:
            cur_n_idx = num_set.index(cur_n)

        for i in range(cur_n_idx + 1, N):
            n = num_set[i]
            p.append(n)
            dfs(lv + 1, n)
            p.pop()
        

N, M = map(int, input().split()) # N: 집합의 원소수, M: 뽑은 개수
num_set = list(map(int, input().split())) 

p = [] # 뽑아 생성한 하나의 순열
num_set.sort()
dfs(0, 0)

File: BOJ/nodejs/15655.js
/*
N과 M (6)
boj.kr/15655
 */
const start = new Date().getTime();

const fs = require("fs");
const input = fs.readFileSync("input.txt").toString().split("\n");
//const input = fs.readFileSync("/dev/stdin").toString().split("\n");

const [N, M] = input[0].split(" ").map(Number);
let nums = input[1].split(" ").map(Number);

nums.sort((a, b) => a - b); // 오름차순 정렬
const check = new Array(N).fill(0);
const S = [];
const dfs = (s) => {
  if (s.length === M) {
    S.push([...s]);

    return;
  }

  for (let i = 0; i < nums.length; i++) {
    // 부분수열이 마지막원소보다 크거나 같을 때만
    let top = s.length - 1;
    if (s.length === 0 || nums[i] >= s[top]) {
      if (check[i] === 0) {
        check[i] = 1;
        s.push(nums[i]);
        dfs(s);
        check[i] = 0;
        s.pop();
      }
    }
  }
};

dfs([]);

let ans = "";
for (let s of S) {
  ans += s.join(" ") + "\n";
}
console.log(ans);

const end = new Date().getTime();

console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
